src/query/ast/src/parser/expr.rs

```diff
@@ -1367,6 +1367,146 @@ pub fn type_name(i: Input) -> IResult<TypeName> {
+            | #ty_int8
+            | #ty_int16
+            | #ty_int32
+            | #ty_int64
+            | #ty_float32
+            | #ty_float64
+            | #ty_decimal
+            | #ty_array
+            | #ty_map
+            | #ty_bitmap
+            | #ty_tuple : "TUPLE(<type>, ...)"
+            | #ty_named_tuple : "TUPLE(<name> <type>, ...)"
+            ) : "type name"
+            },
+            rule! {
+            ( #ty_date
+            | #ty_datetime
+            | #ty_string
+            | #ty_variant
+            | #ty_nullable
+            ) : "type name" },
+        )),
+        |ty| ty,
+    )(i)
+}
+
+pub fn type_name(i: Input) -> IResult<TypeName> {
+    let ty_boolean = value(TypeName::Boolean, rule! { BOOLEAN | BOOL });
+    let ty_uint8 = value(
+        TypeName::UInt8,
+        rule! { ( UINT8 | #map(rule! { TINYINT ~ UNSIGNED }, |(t, _)| t) ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )?  },
+    );
+    let ty_uint16 = value(
+        TypeName::UInt16,
+        rule! { ( UINT16 | #map(rule! { SMALLINT ~ UNSIGNED }, |(t, _)| t) ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_uint32 = value(
+        TypeName::UInt32,
+        rule! { ( UINT32 | #map(rule! { ( INT | INTEGER ) ~ UNSIGNED }, |(t, _)| t) ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_uint64 = value(
+        TypeName::UInt64,
+        rule! { ( UINT64 | UNSIGNED | #map(rule! { BIGINT ~ UNSIGNED }, |(t, _)| t) ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_int8 = value(
+        TypeName::Int8,
+        rule! { ( INT8 | TINYINT ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_int16 = value(
+        TypeName::Int16,
+        rule! { ( INT16 | SMALLINT ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_int32 = value(
+        TypeName::Int32,
+        rule! { ( INT32 | INT | INTEGER ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_int64 = value(
+        TypeName::Int64,
+        rule! { ( INT64 | SIGNED | BIGINT ) ~ ( "(" ~ ^#literal_u64 ~ ^")" )? },
+    );
+    let ty_float32 = value(TypeName::Float32, rule! { FLOAT32 | FLOAT });
+    let ty_float64 = value(
+        TypeName::Float64,
+        rule! { (FLOAT64 | DOUBLE)  ~ PRECISION? },
+    );
+    let ty_decimal = map_res(
+        rule! { DECIMAL ~ "(" ~ #literal_u64 ~ ( "," ~ ^#literal_u64 )? ~ ")" },
+        |(_, _, precision, opt_scale, _)| {
+            Ok(TypeName::Decimal {
+                precision: precision
+                    .try_into()
+                    .map_err(|_| ErrorKind::Other("precision is too large"))?,
+                scale: if let Some((_, scale)) = opt_scale {
+                    scale
+                        .try_into()
+                        .map_err(|_| ErrorKind::Other("scale is too large"))?
+                } else {
+                    0
+                },
+            })
+        },
+    );
+    let ty_array = map(
+        rule! { ARRAY ~ "(" ~ #basic_type_name ~ NOT? ~ NULL? ~ ")" },
```

Comment:
There's too much duplicate code to write this way, and it doesn't handle multiple nested cases and other nested types like map and tuple, for example:
```sql
create table t(a int, b array(array(int not null)));
create table t(a int, b tuple(int not null, string not null));
```
You can write like this example:
```rust
    map_res(
        alt((
            rule! {
            ( #ty_boolean
            | #ty_uint8
            ...
            | #ty_string
            | #ty_variant
            | #ty_nullable
            ) ~ NOT? ~ NULL? : "type name" },
        )),
        |(ty, opt_not, opt_null)| {
            match (opt_not, opt_null) {
                (None, Some(_)) => Ok(ty.wrap_nullable()),
                (Some(_), Some(_)) => Ok(ty),
                (None, None) => Ok(ty),
                (Some(_), None) => Err(ErrorKind::Other("invalid type name")),
            }
        },
    )(i)
```